File: python-cli/workshop/commands/checkin.py
"""workshop/commands/checkin.py — report task completion to the live leaderboard."""

import click
import httpx
from rich import print as rprint

from workshop.utils.agent import is_agent, emit_event, emit_error, emit_complete
from workshop.utils.prompts import agent_text, agent_select
from workshop.utils.config import get_leaderboard_url, read_config

VALID_TASKS = [
    {"id": "todo_1",            "label": "TODO-1 — Add emit_prompt()"},
    {"id": "todo_2",            "label": "TODO-2 — Add is_agent() + NoopSpinner"},
    {"id": "todo_3",            "label": "TODO-3 — Add FlowState"},
    {"id": "todo_4",            "label": "TODO-4 — Add emit_complete()"},
    {"id": "eval_pass",         "label": "Evals ✓ — 7/7 tests passing"},
    {"id": "quest_scaffold",    "label": "quest:scaffold — game scaffolded"},
    {"id": "quest_leaderboard", "label": "quest:leaderboard — lb scaffolded"},
    {"id": "quest_play",        "label": "quest:play — played a round"},
]


def _post_checkin(name: str, task: str, source: str = "manual") -> dict | None:
    url = get_leaderboard_url()
    try:
        r = httpx.post(f"{url}/api/checkin",
                       json={"name": name, "task": task, "source": source},
                       timeout=4.0)
        return r.json()
    except Exception:
        return None


def _post_batch(name: str, tasks: list[str], source: str = "eval") -> dict | None:
    url = get_leaderboard_url()
    try:
        r = httpx.post(f"{url}/api/checkin/batch",
                       json={"name": name, "tasks": tasks, "source": source},
                       timeout=4.0)
        return r.json()
    except Exception:
        return None


@click.group()
def checkin():
    """Check in completed tasks to the live leaderboard."""
    pass


@checkin.command("task")
@click.option("--name",  default=None, help="Your name")
@click.option("--task",  default=None, help="Task ID to check in")
@click.option("--batch", default=None, help="Comma-separated task IDs for batch checkin")
@click.option("--source", default="manual", help="Source: manual | eval | cli")
def checkin_task(name, task, batch, source):
    """Report a completed task to the live leaderboard."""
    agent = is_agent()
    cfg = read_config()

    # Use saved name from config if not provided
    if not name:
        name = cfg.get("workshop_name")
    if not name:
        name = agent_text(
            message="Your name", default="Attendee",
            step=1, of=2, field="name",
        )
    if not name:
        return

    # Batch mode
    if batch:
        tasks = [t.strip() for t in batch.split(",") if t.strip()]
        result = _post_batch(name, tasks, source)
        if result:
            emit_event({"event": "batch_checkin_result", "name": name,
                        "tasks": tasks, "score": result.get("score"),
                        "new_tasks": result.get("new_tasks", [])})
            if not agent:
                rprint(f"[green]✓ Checked in {len(tasks)} tasks for {name} "
                       f"({result.get('score', '?')} pts)[/green]")
        else:
            _lb_unavailable(agent)
        emit_complete([], [])
        return

    # Single task
    if not task:
        task = agent_select(
            message="Which task did you complete?",
            choices=[t["id"] for t in VALID_TASKS],
            default="todo_1",
            step=2, of=2, field="task",
        )
    if not task:
        return

    result = _post_checkin(name, task, source)
    if result:
        emit_event({"event": "checkin_result", "name": name, "task": task,
                    "score": result.get("score"),
                    "total_tasks": result.get("total_tasks")})
        if not agent:
            rprint(f"[green]✓ Checked in: {task} for {name} "
                   f"({result.get('score', '?')} pts)[/green]")
    else:
        _lb_unavailable(agent)
    emit_complete([], [])


def _lb_unavailable(agent: bool):
    emit_error("LEADERBOARD_UNAVAILABLE",
               "Leaderboard not running. Start: cd leaderboard && node server.js", False)
    if not agent:
        rprint("[yellow]Leaderboard not running — checkin not recorded.[/yellow]")
        rprint("[dim]Start it: cd leaderboard && node server.js[/dim]")

File: node-cli/src/commands/workshop/checkin.js
'use strict';

/**
 * workshop:checkin — report task completion to the live leaderboard
 *
 * Auto-called by the eval harness when tests pass.
 * Can also be run manually to check in for any task.
 *
 * Usage:
 *   node src/index.js workshop:checkin
 *   node src/index.js workshop:checkin --name "Alice" --task todo_1
 *   node src/index.js workshop:checkin --name "Alice" --task eval_pass --source eval
 */

const { intro, outro, cancel } = require('@clack/prompts');
const chalk = require('chalk');
const { isAgent, emitEvent, emitComplete, emitError } = require('../../utils/agent');
const { agentText, agentSelect } = require('../../utils/prompt');
const { getLeaderboardUrl } = require('../../utils/config');

const LEADERBOARD = getLeaderboardUrl();

const VALID_TASKS = [
  { id: 'mission_1',   label: 'Mission 1 — Make it speak'           },
  { id: 'mission_2',   label: 'Mission 2 — Make it remember'        },
  { id: 'mission_3',   label: 'Mission 3 — Make it finish'          },
  { id: 'eval_pass',   label: 'Evals ✓ — all tests passing'         },
  { id: 'ipc_channel', label: 'IPC ✓ — TPOK < 500 ms'              },
  { id: 'tmux_launch', label: 'Cockpit — launched via tmux'         },
  { id: 'bonus_cli',          label: 'Bonus CLI — drove a second CLI'              },
  { id: 'bonus_paypal_creds',  label: 'PAY-1 — credentials + prompt events'         },
  { id: 'bonus_paypal_button', label: 'PAY-2 — generated paypal-checkout.html'      },
  { id: 'bonus_paypal_eval',   label: 'PayPal Evals — all bonus tests pass'          },
];

async function workshopCheckin(program) {
  program
    .command('workshop:checkin')
    .description('Check in a completed task to the live leaderboard')
    .option('--name <name>',   'Your name (skips prompt)')
    .option('--task <task>',   'Task ID to check in (skips prompt)')
    .option('--source <src>',  'Source: manual | eval | cli', 'manual')
    .option('--batch <tasks>', 'Comma-separated task IDs for batch checkin')
    .action(async (opts) => {
      const agent = isAgent();
      // Pull saved name from config if not passed explicitly
      const cfg = require('../../utils/config').readConfigFile();
      if (!opts.name && cfg.workshop_name) opts.name = cfg.workshop_name;
      if (!agent) intro(chalk.bold.cyan('✓  Workshop Checkin'));

      // ── Name ──────────────────────────────────────────────────
      let name = opts.name;
      if (!name) {
        name = await agentText({
          message: 'Your name', defaultValue: 'Attendee',
          step: 1, of: 2, field: 'name',
        });
        if (!name) { cancel('Cancelled'); return; }
      }

      // ── Task ──────────────────────────────────────────────────
      // Batch mode (used by eval harness)
      if (opts.batch) {
        const tasks = opts.batch.split(',').map((t) => t.trim()).filter(Boolean);
        await batchCheckin(name, tasks, opts.source);
        if (!agent) outro(chalk.green(`✓ Checked in ${tasks.length} tasks for ${name}`));
        emitComplete([], ['node src/index.js workshop:checkin --name "' + name + '" --task eval_pass']);
        return;
      }

      let task = opts.task;
      if (!task) {
        // No --task flag: just register the name with no task (0 pts for joining).
        // Tasks are reported automatically by evals or with --task flag.
        await joinOnly(name, agent);
        if (!agent) outro(chalk.green(`✓ ${name} registered! Tasks are auto-reported when you run the eval.`));
        emitComplete([], ['node src/index.js workshop:checkin --name "' + name + '" --task eval_pass']);
        return;
      }

      await singleCheckin(name, task, opts.source, agent);
      if (!agent) outro(chalk.green(`✓ Checked in: ${task} for ${name}`));
      emitComplete([], []);
    });
}

async function joinOnly(name, agent) {
  try {
    // POST a batch with empty tasks array — this registers the attendee with 0 pts
    const res = await fetch(`${LEADERBOARD}/api/checkin/batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name, tasks: [], source: 'manual' }),
    });
    const data = await res.json();
    emitEvent({ event: 'checkin_result', name, task: null, score: data.score,
                total_tasks: data.total_tasks, joined: true });
    return data;
  } catch {
    emitError('LEADERBOARD_UNAVAILABLE',
      'Leaderboard not running. Start with: cd leaderboard && node server.js', false);
    if (!agent) console.log(chalk.yellow('\n  Leaderboard not running — your name was not registered.\n  Start it: cd leaderboard && node server.js'));
  }
}

async function singleCheckin(name, task, source, agent) {
  try {
    const res = await fetch(`${LEADERBOARD}/api/checkin`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name, task, source }),
    });
    const data = await res.json();
    emitEvent({ event: 'checkin_result', name, task, score: data.score,
                total_tasks: data.total_tasks, already: data.already });
    return data;
  } catch {
    emitError('LEADERBOARD_UNAVAILABLE',
      'Leaderboard not running. Start with: cd leaderboard && node server.js', false);
    if (!agent) console.log(chalk.yellow('\n  Leaderboard not running — checkin not recorded.\n  Start it: cd leaderboard && node server.js'));
  }
}

async function batchCheckin(name, tasks, source) {
  try {
    const res = await fetch(`${LEADERBOARD}/api/checkin/batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name, tasks, source }),
    });
    const data = await res.json();
    emitEvent({ event: 'batch_checkin_result', name, tasks, score: data.score,
                new_tasks: data.new_tasks, total_tasks: data.total_tasks });
    return data;
  } catch {
    emitError('LEADERBOARD_UNAVAILABLE',
      'Leaderboard not running. Start with: cd leaderboard && node server.js', false);
  }
}

// Export the helpers so the eval harness can call them directly
module.exports = { workshopCheckin, singleCheckin, batchCheckin };
